### DISCUSSION TITLE
How to use ws or wss protocol in Mqtt

### DISCUSSION THREAD

**Discussion by joko2020** [Suraj151/pdi-framework#7] [Q&A]
Hello sir, this **is great library** that i found inside arduino IDE, it's **very useful and nice GUI**, but I want to know more about MQTT for communication over websocket or secure websocket. I read documentation but i cannot find instruction to configure to use WS or WSS, Is It capable to doing Mqtt over Websocket?

Thanks

**Comment by Suraj151** [comment]
Hi @joko2020

the mqtt library used in this framework is using WiFiClient library to send mqtt packets over tcp. its not using any websocket protocol to send the packets.

yes its possible to have mqtt over websocket. but ideally not needed, since i believe websocket library is build to have bidirectional communication between web apps and servers as browsers dont have raw tcp communication api to use.

still if you need this you may need to send the mqtt packets usign websocket client library. you just need to modify send & receive packet functions of this library to use websocket client library instead of wificlient library.

**Comment by joko2020** [comment]
> Hi @joko2020
> 
> the mqtt library used in this framework is using WiFiClient library to send mqtt packets over tcp. its not using any websocket protocol to send the packets.
> 
> yes its possible to have mqtt over websocket. but ideally not needed, since i believe websocket library is build to have bidirectional communication between web apps and servers as browsers dont have raw tcp communication api to use.
> 
> still if you need this you may need to send the mqtt packets usign websocket client library. you just need to modify send & receive packet functions of this library to use websocket client library instead of wificlient library.

@Suraj151 thanks for fast respond , what I'm to do is , I would like to esp8266 to able to communicate in bidirectional to my own cloud server broker in real time.
I setup my Broker server only receive WS & WSS protocol using EMQX server .
My plan is using Your Library for Initial setup instead using WIFIMANAGER LIBRARY for easy access and store credential wifi and mqtt .

**Comment by joko2020** [comment]
Could you give me advice to do

> still if you need this you may need to send the mqtt packets usign websocket client library. you just need to modify send & receive packet functions of this library to use websocket client library instead of wificlient library.

Could You give me advise how to do or give me example sketch to able to use MQTT over websocket

**Comment by Suraj151** [comment]
@joko2020 

since i am not touch base with hardware to test example. i will suggest to use any available esp8266 websocket client library which can send * receive data. and modify send & receive packet functions in mqtt library accordingly.

i dont know exactly when, but if possible i will add option to use websocket for mqtt client.

**Comment by joko2020** [comment]
> @joko2020
> 
> since i am not touch base with hardware to test example. i will suggest to use any available esp8266 websocket client library which can send * receive data. and modify send & receive packet functions in mqtt library accordingly.
> 
> i dont know exactly when, but if possible i will add option to use websocket for mqtt client.

Ok thanks for your Time & advise, hope next version of your framework soon available (👍🏻
